### PR TITLE
lxd/containers: Fix race condition in shutdown

### LINKED ISSUE #4102: LXD Shutdown State Consideration
When viewing the code for containerShutdown, it seems there may be a small bug in relation to when the container's state is checked during lxd shutdown.

First the container's state is cached during shutdown:
https://github.com/lxc/lxd/blob/2d9eb5ddbadd87d80ea8131c3d86bdcb54a2416c/lxd/containers.go#L213

However, when the code actually considers whether or not the container should be shutdown, it does not use the cached copy of the state, preferring to request if the container is currently running:
https://github.com/lxc/lxd/blob/2d9eb5ddbadd87d80ea8131c3d86bdcb54a2416c/lxd/containers.go#L224

Ergo, when the container is either shutdown or left alone in the below lines, the data stored to "volatile.last_state.power" would seem to not be the last state considered by the shutdown code as it uses the cached copy of the state stored before checking IsRunning:
https://github.com/lxc/lxd/blob/2d9eb5ddbadd87d80ea8131c3d86bdcb54a2416c/lxd/containers.go#L239
https://github.com/lxc/lxd/blob/2d9eb5ddbadd87d80ea8131c3d86bdcb54a2416c/lxd/containers.go#L244

For reference this is the for loop I am referring to below:
https://github.com/lxc/lxd/blob/2d9eb5ddbadd87d80ea8131c3d86bdcb54a2416c/lxd/containers.go#L209

If this is indeed an issue, a possible fix would be to utilize the cached state variable instead of IsRunning.  Additionally, we could also consider caching state prior to the for loop in order to minimize the time that the code checks state in order to try to make lxd shutdown's behavior more predictable to the user because, if there were to be any long pause between caching the var and checking IsRunning, behavior may not be what is expected when the user issued the lxd shutdown command.

### PR DESCRIPTION
**Pull request by stgraber**
Closes #4102

Signed-off-by: Stéphane Graber <stgraber@ubuntu.com>

### CODE REVIEW

File: lxd/containers.go [hunk 1]
```diff
@@ -220,8 +217,11 @@ func containersShutdown(s *state.State) error {
 			wg.Wait()
 		}
 
+		// Record the current state
+		lastState := c.State()
+
 		// Stop the container
-		if c.IsRunning() {
+		if lastState != "BROKEN" && lastState != "STOPPED" {
```
Comment (brauner) [inline lxd/containers.go]: I don't know where the `BROKEN` state is coming from `liblxc` doesn't know this state iirc and `go-lxc` doesn't define it either. Did you mean `ABORTING` or `STOPPING`?

Comment (stgraber) [reply lxd/containers.go]: It's a LXD state.

Comment (stgraber) [reply lxd/containers.go]: That's what we return if we timeout getting a state from liblxc and likely what we'll be using when a cluster node is down.

**Comment by stgraber** [comment]
@brauner should be good to go

**Pull request merged**